[1/N][C10D] Add a customized ScubaLogHandler implementation for internal FB use (#86699)

wz337 opened the pull request:
Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/86699

This diff does the following:
1. **c10d_error_logger.py**: Add an API  to create a logger with a specific logging handler based on the destination.
2. The API from above would get a logging handler based on the destination provided.
-  **caffe2/torch/distributed/logging_handlers.py**: For OSS, we simply use a NullHandler() for now.
3. Add associated test files for 1 and 2.


Test Plan:
## Unit Test
```
buck test @//mode/dev-nosan //caffe2/test/distributed:test_c10d_error_logger -- --print-passing-details
```
```
File changed: fbcode//caffe2/test/distributed/test_c10d_error_logger.py
File changed: fbsource//xplat/caffe2/test/distributed/TARGETS
9 additional file changes
waiting for all tests to finish...
✓ Listing success: caffe2/test/distributed:test_c10d_error_logger (0.2s)
Found 1 tests
✓ Pass: caffe2/test/distributed:test_c10d_error_logger - test_get_or_create_logger (caffe2.test.distributed.test_c10d_error_logger.C10dErrorLoggerTest) (0.2s)

stdout:

stderr:

Buck UI:      https://www.internalfb.com/buck2/b975f6b0-77e9-4287-8722-f95b48036181
Test Session: https://www.internalfb.com/intern/testinfra/testrun/1407375150206593
RE: reSessionID-4d7ab8ca-1051-48e9-a5a8-6edbe15d1fe4  Up: 124 B  Down: 0 B
Jobs completed: 5. Time elapsed: 3.5s.
Tests finished: Pass 1. Fail 0. Fatal 0. Skip 0. 0 builds failed
```

Differential Revision: D39920391

rohan-varma reviewed (commented):
Do we need C++-side support as well?

wz337 commented:
/easycla

wz337 commented:
> Do we need C++-side support as well?

Are you referring to the underlying c++ trace? If it is, then the error also records this information.

wz337 commented:
/easycla

wz337 commented:
/easycla

fduwjj reviewed (approved):
Accept since the diff has been accepted on fbcode.

H-Huang reviewed (approved):
LGTM!

wz337 committed:
[1/N][C10D] Add a customized ScubaLogHandler implementation for internal FB use (#87123)

Reviewed By: H-Huang, fduwjj

Differential Revision: D39920391

fbshipit-source-id: 4d2fd3970e981153588cafef484af3d63d6f641a
